### DISCUSSION TITLE
Why renovate still try to update dependencies in subofolders?

### DISCUSSION THREAD

**Discussion by gremo** [renovatebot/renovate#9465] [Request Help] [answered]
With the following configuration, Renovate still open PRs to "Pin dependencies"... what I'm doing wrong?

```json
{
  "extends": [
    "config:base",
    ":disablePeerDependencies"
  ],
  "packageRules": [
    {
      "matchPaths": ["+(package.json)"],
      "depTypeList": [
        "devDependencies"
      ],
      "rangeStrategy": "bump"
    }
  ],
  "semanticCommits": "disabled"
}
```

Test repository: https://github.com/gremo/test-renovate

**Comment by viceice** [comment, marked as answer]
because you did not disabled it. you only changed `rangeStrategy` for the root `package.json`

**Reply by gremo** [reply]
Thanks. So maybe I need to disable it globally and then enable it for the root package.json, like this way?

```json
{
  "extends": [
    "config:base",
    ":disablePeerDependencies"
  ],
  "packageRules": [
    {
      "matchPaths": ["package.json"],
      "enabled": false
    },
    {
      "matchPaths": ["+(package.json)"],
      "depTypeList": [
        "devDependencies"
      ],
      "rangeStrategy": "bump"
    }
  ],
  "semanticCommits": "disabled"
}
```

**Reply by rarkins** [reply]
That ain't work because you're disabling all package.json in the first rule and then not re-enabling anything later. 

Have one package rule to disable all. Then another with matchFiles=package.json and enabled=true to re-enable just that one. 

You haven't described what you're hoping to achieve with the other fields you added so I can't recommend about them yet

**Reply by gremo** [reply]
I feel dumb every time I try to configure Renovate. Every single thing I try doesn't work 😢 of course, because of me. 

Anyways I'm trying to do a simple thing:
1. Ignore anything but root `package.json` (I assume `"enabled": true` in the second rule as you said?)
2. Ignore peer dependencies
3. Pin dependencies and automerge only patch versions
4. Automerge devDependecies (bump)


Even the simple automerge doesn't work for me, Renovate still open PRs.

```json5
{
  "extends": [
    "config:base",
    ":disablePeerDependencies" // 2.
  ],
  "semanticCommits": "disabled",
  "packageRules": [
    {
      "matchPaths": ["package.json"], // 1.
      "enabled": false
    },
    {
      "matchDepTypes": [ // 3.
        "dependencies"
      ],
      "matchUpdateTypes": [
        "minor",
        "patch"
      ],
      "enabled": true,
      "rangeStrategy": "pin",
      "automerge": true
    },
    {
      "matchDepTypes": [ // 4.
        "devDependencies"
      ],
      "enabled": true,
      "rangeStrategy": "bump",
      "automerge": true
    }
  ]
}
```

**Reply by rarkins** [reply]
If in doubt, put everything in a package rule. Remember:
- Package rules are matched and then applied in order. None are skipped even if a previous one matched
- Every package rule must have at least one match* or exclude* field for matching
- Every package rule must have at least one non-match and non-exclude option to apply

Your rules do not follow my previous recommendations, so I suggest you start with that first:

> Have one package rule to disable all. Then another with matchFiles=package.json and enabled=true to re-enable just that one.

**Reply by rarkins** [reply]
Also, what you are trying to do is not "simple". You are starting out with complex and specific requirements and normally I'd recommend you start with defaults and see how that goes.

**Reply by gremo** [reply]
Thank you @rarkins for your time and patience, I'll follow your suggestion. Right now I've read the documentation again, and there are some aspects hard to understand (like `:skipStatusChecks`, I was not aware).

Anyways I ended up with this: https://github.com/gremo/test-renovate/blob/master/renovate.json which "partially works", but probably I'll open an issue about that.  All these dependencies where "outdated", but for some reason Renovate **opened and automatically merged only updates for `classnames` and `eslint`**.

Starting `package.json`:

```json
{
  "name": "test-renovate",
  "private": true,
  "version": "1.0.0",
  "license": "ISC",
  "dependencies": {
    "classnames": "2.3.0",
    "react": "17.0.1",
    "react-dom": "17.0.1"
  },
  "devDependencies": {
    "eslint": "^7.22.0",
    "stylelint": "^13.11.0"
  }
}
```

`renovate.json`:

```
{
  "extends": [
    "config:base",
    ":disablePeerDependencies",
    ":semanticCommitsDisabled",
    ":skipStatusChecks"
  ],
  "packageRules": [
    {
      "matchDepTypes": ["dependencies"],
      "matchUpdateTypes": [
        "minor",
        "patch",
        "pin",
        "digest"
      ],
      "rangeStrategy": "pin",
      "automerge": true
    },
    {
      "matchDepTypes": ["devDependencies"],
      "rangeStrategy": "bump",
      "automerge": true
    }
  ]
}
```

After (note that `react` `react-dom` and `stylelint` where not updated):

```json
{
  "name": "test-renovate",
  "private": true,
  "version": "1.0.0",
  "license": "ISC",
  "dependencies": {
    "classnames": "2.3.1",
    "react": "17.0.1",
    "react-dom": "17.0.1"
  },
  "devDependencies": {
    "eslint": "^7.23.0",
    "stylelint": "^13.11.0"
  }
}
```

**Reply by viceice** [reply]
```json
{
      "matchDepTypes": ["devDependencies"],
      "rangeStrategy": "bump",
      "automerge": true
    }
```

This will only bump if the current range doen't match the new version. try `replace` strategy.

**Reply by viceice** [reply]
**WARNING** You should not skip status checks with automerge, it can break your main branch if there is something wrong with the update!

**Reply by gremo** [reply]
@viceice thanks, it turns out that Renovate actually updated all dependencies as expected, but `react` `react-dom` and `stylelint` where updated 3 hours later than `classnames` and eslint. Don't know why.

**Reply by gremo** [reply]
> **WARNING** You should not skip status checks with automerge, it can break your main branch if there is something wrong with the update!

I know :) it's a testing repository. When I use Renovate in production, I always have some kind of testing.

**Comment by gremo** [comment]
Still struggling into this. can't make it work. I would say "disable renovate globally and then enable only for the root `package.json`). I guess I still can't understand how renovate evaluate the configuration... thanks

```json
{
  "extends": [
    "config:base",
    ":automergeMinor",
    ":automergePatch",
    ":pinDevDependencies",
    ":semanticCommitsDisabled",
    ":skipStatusChecks",
    "::disableRenovate"
  ],
  "packageRules": [
    {
      "matchFiles": ["package.json"],
      "enabled": true
    },
    {
      "matchDepTypes": ["dependencies"],
      "rangeStrategy": "bump"
    }
  ]
}
```

**Reply by viceice** [reply]
You have a typo ` ::disableRenovate` in your preset.
 also i think using `:disableRenovate` preset will completly disable renovate.

**Reply by rarkins** [reply]
```
{
  "extends": [
    "config:base",
    ":automergeMinor",
    ":automergePatch",
    ":pinDevDependencies",
    ":semanticCommitsDisabled",
    ":skipStatusChecks"
  ],
  "enabledManagers": ["npm"],
  "packageRules": [
    {
      "matchPackagePatterns": ["*"],
      "enabled": false
    },
    {
      "matchFiles": ["package.json"],
      "enabled": true
    },
    {
      "matchDepTypes": ["dependencies"],
      "rangeStrategy": "bump"
    }
  ]
}
```

**Reply by rarkins** [reply]
If in doubt, always use `packageRules`

**Reply by gremo** [reply]
> You have a typo ` ::disableRenovate` in your preset.
> also i think using `:disableRenovate` preset will completly disable renovate.

Oh you right, I'm going to test it without the typo 👍 

**EDIT**: it doesn't work

**Reply by gremo** [reply]
> ```
> {
>   "extends": [
>     "config:base",
>     ":automergeMinor",
>     ":automergePatch",
>     ":pinDevDependencies",
>     ":semanticCommitsDisabled",
>     ":skipStatusChecks"
>   ],
>   "enabledManagers": ["npm"],
>   "packageRules": [
>     {
>       "matchPackagePatterns": ["*"],
>       "enabled": false
>     },
>     {
>       "matchFiles": ["package.json"],
>       "enabled": true
>     },
>     {
>       "matchDepTypes": ["dependencies"],
>       "rangeStrategy": "bump"
>     }
>   ]
> }
> ```

Thanks, going to test it. So verbose!

**EDIT**: works OK thanks, I changed the fist package rule to `matchPaths `(seems more appropriate)
